java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,164 @@
+    .ofTypes("java.time.format.DateTimeFormatterBuilder")
+    .names("appendValue")
+    .addParametersMatcher("java.time.temporal.TemporalField", "int")
+    .build();
+
+  private static final MethodMatchers WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekBasedYear")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final MethodMatchers WEEK_OF_WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekOfWeekBasedYear")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final String CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE = "Change this year format to use the week-based year instead.";
+  private static final String CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE = "Change this year format to use ChronoField.YEAR instead.";
+  private static final String SECONDARY_LOCATION_MESSAGE = "";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree invocation = (MethodInvocationTree) tree;
+    if (APPEND_VALUE_MATCHER.matches(invocation)) {
+      ChainedInvocationVisitor visitor = new ChainedInvocationVisitor();
+      invocation.accept(visitor);
+      if (!visitor.usesWeek || !visitor.usesYear) {
+        return;
+      }
+      if (visitor.usesWeekBasedYear && !visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE, visitor.secondaries, null);
+      } else if (!visitor.usesWeekBasedYear && visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE, visitor.secondaries, null);
+      }
+    }
+  }
+
+  private static class ChainedInvocationVisitor extends BaseTreeVisitor {
+    private boolean usesWeek = false;
+    private boolean usesWeekOfWeekBasedYear = false;
+    private boolean usesYear = false;
+    private boolean usesWeekBasedYear = false;
+    private final List<JavaFileScannerContext.Location> secondaries = new ArrayList<>();
+    private ExpressionTree primary = null;
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree callToAppendValue) {
+      MethodInvocationTree tree = callToAppendValue;
+      inspectCall(tree);
+      Optional<MethodInvocationTree> current = MethodTreeUtils.subsequentMethodInvocation(tree, APPEND_VALUE_MATCHER);
+      while (current.isPresent()) {
+        tree = current.get();
+        inspectCall(tree);
+        current = MethodTreeUtils.subsequentMethodInvocation(tree, APPEND_VALUE_MATCHER);
+      }
+    }
+
+    private void inspectCall(MethodInvocationTree invocation) {
+      ExpressionTree argument = invocation.arguments().get(0);
+      if (refersToYear(argument)) {
+        usesYear = true;
+        primary = argument;
```

Comment:
If there are multiple different years in the formatter, we can end up in a situation where `usesWeekBasedYear` is true, but `primary` is not a week based year.

Reply:
Done